### PR TITLE
Secrets Interface

### PR DESCRIPTION
**Pull request by srkukarni**
### Motivation

Many a times functions/sources/sinks need access to secrets like passwords in a secure way. They really cannot be in code nor in config thats passed to the function.
This pr adds a secrets interface to the functions api. It also adds a SecretsProvider interface that actually fetches the secrets. Implementation of the SecretProvider interface will fetch the secret from the actual underlying secret store. This pr also adds a default environment based secrets gathering using kubernetes secrets.

### Modifications

Describe the modifications you've done.

### Result

After your change, what will change.

### CODE REVIEW

File: pulsar-functions/instance/src/main/python/python_instance_main.py [hunk 1]
```diff
@@ -72,6 +74,8 @@ def main():
   parser.add_argument('--logging_file', required=True, help='Log file name')
   parser.add_argument('--expected_healthcheck_interval', required=True, help='Expected time in seconds between health checks', type=int)
   parser.add_argument('--install_usercode_dependencies', required=False, help='For packaged python like wheel files, do we need to install all dependencies', type=bool)
+  parser.add_argument('--secrets_provider', required=True, help='The classname of the secrets provider')
```
Comment (cckellogg) [inline pulsar-functions/instance/src/main/python/python_instance_main.py]: Should the secrets be optional flags?

File: pulsar-common/src/main/java/org/apache/pulsar/common/io/SinkConfig.java [hunk 2]
```diff
@@ -55,6 +55,7 @@
     private Map<String, ConsumerConfig> inputSpecs = new TreeMap<>();
 
     private Map<String, Object> configs;
+    private Map<String, String> secrets;
```
[the remark this answers is not in the document]
Comment (srkukarni) [reply pulsar-common/src/main/java/org/apache/pulsar/common/io/SinkConfig.java]: Isn't this already two levels. SecretName and SecretPath?

File: pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/KubernetesRuntime.java [hunk 3]
```diff
@@ -519,12 +535,27 @@ private V1Container getFunctionContainer(List<String> instanceCommand, Function.
...
         // setup the environment variables for the container
+        List<V1EnvVar> envVars = new LinkedList<>();
         final V1EnvVar envVarPodName = new V1EnvVar();
         envVarPodName.name("POD_NAME")
                 .valueFrom(new V1EnvVarSource()
                         .fieldRef(new V1ObjectFieldSelector()
                                 .fieldPath("metadata.name")));
-        container.setEnv(Arrays.asList(envVarPodName));
+        envVars.add(envVarPodName);
+        if (!StringUtils.isEmpty(instanceConfig.getFunctionDetails().getSecretsMap())) {
```
Comment (cckellogg) [inline pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/KubernetesRuntime.java]: can this abstracted and delegated to a secrets backend provider

File: pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/RuntimeUtils.java [hunk 4]
```diff
@@ -136,6 +139,13 @@
         }
         args.add("--expected_healthcheck_interval");
         args.add(String.valueOf(expectedHealthCheckInterval));
+
+        args.add("--secrets_provider");
```
Comment (cckellogg) [inline pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/RuntimeUtils.java]: should these be optional?

File: pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdSinks.java [hunk 5]
```diff
@@ -406,6 +408,12 @@ void processArguments() throws Exception {
                 sinkConfig.setConfigs(parseConfigs(sinkConfigString));
             }
 
+            if (null != secretsConfig) {
```
Comment (cckellogg) [inline pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdSinks.java]: to start maybe secrets can only be defined in the function config or user config. secrets might need more than just a single key=value

File: pulsar-common/src/main/java/org/apache/pulsar/common/io/SinkConfig.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (cckellogg) [reply pulsar-common/src/main/java/org/apache/pulsar/common/io/SinkConfig.java]: +1

**Comment by srkukarni** [comment]
rerun integration tests

**Comment by srkukarni** [comment]
@merlimat @cckellogg I've made the changes as requested. Please take a look again

File: pulsar-functions/instance/src/main/python/secretsprovider.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+#
+
+# -*- encoding: utf-8 -*-
+
+"""secretsprovider.py: Interfaces and definitions for Secret Providers
+"""
+from abc import abstractmethod
+import os
+
+class SecretsProvider:
```
[the remark this answers is not in the document]
Comment (cckellogg) [reply pulsar-functions/instance/src/main/python/secretsprovider.py]: I think this ok now i think about it more. More implementations can be added later

**Comment by srkukarni** [comment]
I will break the change into smaller changes and it might be best to comment on each individual pr

File: pulsar-functions/secretsprovider/src/main/java/org/apache/pulsar/functions/secretsproviderconfigurator/KubernetesSecretsProviderConfigurator.java [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    public Map<String, String> getSecretsProviderConfig(Function.FunctionDetails functionDetails) {
+        return null;
+    }
+
+    @Override
+    public void configureKubernetesRuntimeSecretsProvider(V1Container container, Function.FunctionDetails functionDetails) {
+        if (!StringUtils.isEmpty(functionDetails.getSecretsMap())) {
+            Type type = new TypeToken<Map<String, Object>>() {
+            }.getType();
+            Map<String, Object> secretsMap = new Gson().fromJson(functionDetails.getSecretsMap(), type);
```
Comment (jerrypeng) [inline pulsar-functions/secretsprovider/src/main/java/org/apache/pulsar/functions/secretsproviderconfigurator/KubernetesSecretsProviderConfigurator.java]: Perhaps its worthwhile to explicitly define the structure of the "Object" for kubernetes so it is clear what configs should be there. You can just probably declare a class that represents the Object

**Comment by merlimat** [comment]
@srkukarni This can be closed?

**Comment by srkukarni** [comment]
@merlimat will close it once #2875 and a follow up example/doc change goes in

**Comment by srkukarni** [comment]
Closing because this has since been subsumed by other prs

**Pull request closed without merging**